### ISSUE TITLE
Runtimes slow down dramatically proportionally with feature file size

### ISSUE THREAD

**Issue by tgsmith61591** [cucumber-rs/cucumber#331]
We maintain a large collection of feature files that feed nightly regression tests, the runtime of which has grown significantly recently. These are generally maintained in logically separated feature files, and leverage `Scenario Outline` tables, sometimes with 100-200 scenarios per feature file.

In experimenting with optimizations on a single tag, I tried splitting a feature file into 4 and observed a massive performance gain. Here is my baseline:

```
[Summary]
1 feature
257 scenarios (226 passed, 31 failed)
1663 steps (1632 passed, 31 failed)
Tests completed in 266 sec
```

Here are the same exact tests divided over 4 feature files:
```
[Summary]
4 features
257 scenarios (226 passed, 31 failed)
1663 steps (1632 passed, 31 failed)
Tests completed in 99 sec
```

For reference,  here is how we're running (note that `local_test_jobs` keeps Bazel from trying to do its own parallelism, instead delegating the concurrency to the cucumber engine):

```bash
$ bazel run //my-crate:my-test-suite \
    --local_test_jobs=1 -- \
    --concurrency 16 \
    --tags="@my-cool-tag"
```

Several questions I have after observing this major performance difference:

- How is concurrency _actually_ affecting the runtime? I was under the impression concurrency was at the `scenario` level, but now I'm wondering whether it's actually at the feature level
- Is there any guidance you can give on tuning `concurrency` to the number of feature files?
- Is there any further guidance on how tests should be broken up to get the best performance out of the `cucumber-rs` engine?

**Comment by tyranron** [comment]
@tgsmith61591 

> * How is concurrency _actually_ affecting the runtime? I was under the impression concurrency was at the `scenario` level, but now I'm wondering whether it's actually at the feature level

`concurrency` is the maximum number of scenarios running concurrently (in `async` manner).

> * Is there any guidance you can give on tuning `concurrency` to the number of feature files?
> * Is there any further guidance on how tests should be broken up to get the best performance out of the `cucumber-rs` engine?

Actually, you shouldn't. The behavior you observe, that splitting source files leads to a significant performance gain, seems to be buggy, weird and unexpected. There shouldn't be any significant difference. Should be investigated and fixed.

The idea we have in mind for now is that a `Parser` returns a `Stream` of features being consumed by a `Runner`, executing them concurrently on scenario level. Seems like the current `Runner` implementation breaks up those features into scenarios in some weird manner, affecting the performance.

**Comment by tyranron** [comment]
@ilslv do you have any suggestions on this?

**Comment by ilslv** [comment]
@tgsmith61591 can you share a little bit more about the characteristics of the testing suite? Is it async or sync heavy, or maybe appropriately both? Can you share the `World` setup: number of `concurrency` and other options?

**Comment by tgsmith61591** [comment]
Hey @ilslv and @tyranron, the test suite is very `async` heavy. While I cannot share the world setup (highly complex, and belongs to the company, not me) I can share a [small repo](https://github.com/tgsmith61591/cucumber-rs-bug) I set up to reproduce this issue.

### tl;dr

* Execute 500 scenarios that sleep for 1 second
* Example A puts 500 scenarios into a single file
* Example B spreads 500 scenarios over 10 files
* Example A takes 5:30 - 6 minutes at `-c 32`
* Example B takes **<20 seconds** at `-c 32` (!!!)

**Comment by ilslv** [comment]
Thank you for the reproduction repo! I'll definitely take a look hopefully this weekend.

**Comment by tgsmith61591** [comment]
Hey @ilslv, any chance you got a chance to look at this?

**Comment by ilslv** [comment]
not yet, unfortunately 😢

**Comment by flyingsilverfin** [comment]
We might be able to take a look into this if you could provide any pointers @ilslv ?  We have 1000 tests or more that currently eat up around 40 minutes instead of around 5 minutes!

**Comment by dmitrii-ubskii** [comment]
We've come up with a simple workaround using a wrapper around the basic parser that splits each scenario into its own gherkin feature:
```rs
#[derive(Debug, Default)]
struct SingletonParser {
    basic: cucumber::parser::Basic,
}

impl<I: AsRef<Path>> cucumber::Parser<I> for SingletonParser {
    type Cli = <cucumber::parser::Basic as cucumber::Parser<I>>::Cli;
    type Output = stream::FlatMap<
        stream::Iter<std::vec::IntoIter<Result<Feature, cucumber::parser::Error>>>,
        Either<
            stream::Iter<std::vec::IntoIter<Result<Feature, cucumber::parser::Error>>>,
            stream::Iter<iter::Once<Result<Feature, cucumber::parser::Error>>>,
        >,
        fn(
            Result<Feature, cucumber::parser::Error>,
        ) -> Either<
            stream::Iter<std::vec::IntoIter<Result<Feature, cucumber::parser::Error>>>,
            stream::Iter<iter::Once<Result<Feature, cucumber::parser::Error>>>,
        >,
    >;

    fn parse(self, input: I, cli: Self::Cli) -> Self::Output {
        self.basic.parse(input, cli).flat_map(|res| match res {
            Ok(mut feature) => {
                let scenarios = mem::take(&mut feature.scenarios);
                let singleton_features = scenarios
                    .into_iter()
                    .map(|scenario| {
                        Ok(Feature {
                            name: feature.name.clone() + " :: " + &scenario.name,
                            scenarios: vec![scenario],
                            ..feature.clone()
                        })
                    })
                    .collect_vec();
                Either::Left(stream::iter(singleton_features))
            }
            Err(err) => Either::Right(stream::iter(iter::once(Err(err)))),
        })
    }
}

```

Before:
```
[Summary]
1 feature
702 scenarios (702 passed)
41957 steps (41957 passed)
test test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2013.41s
```

After:
```
[Summary]
702 features
702 scenarios (702 passed)
41957 steps (41957 passed)
test test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 165.86s
```

**Comment by tyranron** [comment]
@tgsmith61591 @dmitrii-ubskii @flyingsilverfin sorry for getting too late on this.

First of all, many thanks to @tgsmith61591 for providing [such a nice repro](https://github.com/tgsmith61591/cucumber-rs-bug), allowing to investigate the bug!

I've tracked down the cause of this issue. The bigger `.feature` file is - the bigger [`gherkin::Feature` structure](https://docs.rs/gherkin/0.14.0/gherkin/struct.Feature.html) is parsed out of it. Once we have it quite big, the usual `Clone` operations become quite expensive. We've already used `Arc` through the codebase for passing them around, but there were still some places doing so unwise. The one I tracked down instanly and fixed in https://github.com/cucumber-rs/cucumber/commit/562e2ba07a391c6e59834fe5edb12f74222cd4a3.

However, this is only a part of the problem. Since `gherkin` types use [naive `#[derive(Hash, PartialEq)]` implementations](https://docs.rs/gherkin/0.14.0/src/gherkin/lib.rs.html#162), they become too expensive for large structures. Framkly, the `cargo-flamegraph` shows on the [repro](https://github.com/tgsmith61591/cucumber-rs-bug) that `<gherkin::Table as Hash>::hash()` ate 42% CPU time and `<gherkin::Feature as PartialEq>::eq()` ate 22% CPU time 😱

![test-monolith](https://github.com/user-attachments/assets/c9ae73d6-1a09-4a82-8da4-3cb113c51777)

The fix is on the way, I'll try to get it merged next week!
